exersices/Ex06/logistic_regression.py

```diff
@@ -0,0 +1,268 @@
+
+
+__author__ = "Sindre Tallaksrud"
+__email__ = "sindrtal@nmbu.no"
+
+
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.exceptions import NotFittedError
+from sklearn.utils import check_random_state, check_X_y
+
+
+def sigmoid(z):
+    """Perform a logistic transform on the input.
+    This function applies the sigmoidal function element-wise to all
+    elements of `z`. The sigmoidal function is on the following form:
+    Parameters
+    ----------
+    z : np.ndarray
+        Logit to transform.
+    Returns
+    -------
+    sigmoidal_transformed_z : np.ndarray
+        Transformed input.
+    """
+    return 1/(1+np.exp(-z))
+
+
+def predict_proba(coef, X):
+    """Predict the class probabilities for each data point in :math:`X`.
+    Estimate which class each data point in X corresponds to. This is done
+    according to the following formula.
+    .. math::
+        \hat{y}_i = \sigma(\mathbf{x}_i^T \mathbf{w}),
+    where :math:`x_i` is the i-th row in :math:`X` and :math:`\sigma` is
+    the sigmoidal function. Alternatively, in matrix-vector form:
+    .. math::
+        \hat{\mathbf{y}} = \sigma(X \mathbf{w}).
+    Parameters
+    ----------
+    coef : np.ndarray(shape=(r,))
+        The weight vector, :math:`w`
+    X : np.ndarray(shape=(n, r))
+        The data matrix (aka design or measurement matrix)
+    Returns
+    -------
+    p : np.ndarray(shape(n,))
+        The predicted class probabilities.
+    """
+    y_hat = sigmoid(X@coef)
+    return y_hat
+
+
+def logistic_gradient(coef, X, y):
+    """Returns the gradient of a logistic regression model.
+    Parameters
+    ----------
+    coef : np.ndarray(shape=(r,))
+        The weight vector, :math:`w`
+    X : np.ndarray(shape=(n, r))
+        The data matrix (aka design or measurement matrix)
+    y : np.ndarray(shape=(n,))
+        The true class labels for each data point.
+    Returns
+    -------
+    gradient : np.ndarray(shape=(r,))
+        The gradient of the cross entropy loss related to the linear
+        logistic regression model.
+    """
+    y_en = predict_proba(coef, X)
+    return -(y - y_en)@X
```

Comment:
This is not wrong, but looks a bit confused. `return (y_en - y) @ X` would be better.